tests/test_helpers/dataset_utils.py

```diff
@@ -0,0 +1,258 @@
+                r"\1 [\2]",
+                yaml.dump(
+                    {"dataset": [generated_dataset]},
+                    default_flow_style=False,
+                    sort_keys=False,
+                    indent=2,
+                ),
+            )
+        )
+
+
+def generate_dataset(
+    existing_dataset: Dict[str, Any],
+    api_data: Dict[str, List[Row]],
+    collection_order: List[str] = None,
+):
+    """
+    Generates a dataset which is an aggregate of the existing dataset and
+    any new fields generated from the API data. Orders the collections
+    based on the order of collection_order.
+    """
+
+    # preserve the collection order in the dataset if a collection order is not provided
+    if not collection_order:
+        collection_order = [
+            collection["name"] for collection in existing_dataset["collections"]
+        ]
+
+    # remove the dataset name from the keys in the api_data map before passing
+    # into generate_collections
+    generated_collections = generate_collections(
+        {
+            collection_name.replace(f"{existing_dataset['fides_key']}:", ""): collection
+            for collection_name, collection in api_data.items()
+        },
+        existing_dataset,
+    )
+
+    return {
+        "fides_key": existing_dataset["fides_key"],
+        "name": existing_dataset["name"],
+        "description": existing_dataset["description"],
+        "collections": [
+            {
+                "name": collection["name"],
+                "fields": collection["fields"],
+            }
+            for collection in sorted(
+                generated_collections,
+                key=lambda collection: collection_order.index(collection["name"]),
+            )
+        ],
+    }
+
+
+def generate_collections(
+    api_data: Dict[str, List[Row]], dataset: Dict[str, Any] = None
+) -> List[Dict[str, Any]]:
+    """
+    Generates a list of collections based on the response data or returns
+    the existing collections if no API data is available.
+    """
+
+    # convert FidesopsDataset to Dataset to be able to use the Collection helpers
+    collection_map = {}
+    if dataset:
+        graph = convert_dataset_to_graph(
+            FidesopsDataset(**dataset), dataset["fides_key"]
+        )
+        collection_map = {
+            collection.name: collection for collection in graph.collections
+        }
+
+    collections = []
+    for collection_name in set().union(
+        api_data.keys(),
+        collection_map.keys(),
+    ):
+        if len(rows := api_data.get(collection_name, [])):
+            fields = generate_fields(rows[0], collection_name, collection_map)
+        else:
+            fields = get_simple_fields(collection_map.get(collection_name).fields)
+
+        collections.append(
+            {
+                "name": collection_name,
+                "fields": fields,
+            }
+        )
+
+    return collections
+
+
+def generate_fields(
```

Comment:
We have a `merge_datasets` function but we need to preserve the order of the fields and we lose control of that as soon as we convert a simple dict to a `FidesopsDataset`, the fields are alphabetized due to default Pydantic functionality.